src/optimizer/join_elimination.cpp

```diff
@@ -0,0 +1,184 @@
+#include "duckdb/planner/operator/logical_order.hpp"
+#include "duckdb/planner/operator/logical_simple.hpp"
+#include "duckdb/storage/table_storage_info.hpp"
+
+namespace duckdb {
+unique_ptr<LogicalOperator> JoinElimination::Optimize(unique_ptr<LogicalOperator> op) {
+	if (!finish_collection) {
+		// We need to collect the unique constraints set at first, and we only need to collect once.
+		CollectUniqueConstraintsSet(*op);
+		finish_collection = true;
+		if (unique_constraints_set.empty()) {
+			// There are no unique constraints set, so we can return directly.
+			return op;
+		}
+	}
+
+	switch (op->type) {
+	case LogicalOperatorType::LOGICAL_COMPARISON_JOIN: {
+		// Only support the normal comparison join now.
+		auto &join = op->Cast<LogicalComparisonJoin>();
+		idx_t inner_child_idx, outer_child_idx;
+		auto join_type = join.join_type;
+		// We only need to handle the left outer join here,
+		// because at this point in the plan, only left joins exist.
+		if (IsLeftOuterJoin(join_type)) {
+			inner_child_idx = 1;
+			outer_child_idx = 0;
+		} else {
+			D_ASSERT(!IsRightOuterJoin(join_type));
+			break;
+		}
+
+		// Check whether there are used columns from the parent nodes are came from the inner side.
+		bool all_matched = true;
+		auto &inner_child = join.children[inner_child_idx];
+		const auto &inner_bindings = inner_child->GetColumnBindings();
+		for (idx_t col_idx = 0; col_idx < inner_bindings.size(); col_idx++) {
+			if (column_references.find(inner_bindings[col_idx]) != column_references.end()) {
+				all_matched = false;
+				break;
+			}
+		}
+		if (!all_matched) {
+			break;
+		}
+
+		auto &outer_child = join.children[outer_child_idx];
+		const auto &outer_bindings = outer_child->GetColumnBindings();
+		column_binding_set_t outer_bindings_set;
+		for (auto &binding : outer_bindings) {
+			outer_bindings_set.insert(binding);
+		}
+		column_binding_set_t inner_bindings_set;
+		for (auto &binding : inner_bindings) {
+			inner_bindings_set.insert(binding);
+		}
+		// Collect the join key from the inner side
+		column_binding_set_t inner_keys_set;
+		// We need to check whether the join conditions are valid or not.
+		// At first, we make more strict check.
+		bool join_conditions_valid = true;
+		for (const auto &cond : join.conditions) {
+			if (cond.comparison != ExpressionType::COMPARE_EQUAL ||
+			    cond.left->type != ExpressionType::BOUND_COLUMN_REF ||
+			    cond.right->type != ExpressionType::BOUND_COLUMN_REF) {
+				join_conditions_valid = false;
+				break;
+			}
+			ColumnBinding &left_column_binding = cond.left->Cast<BoundColumnRefExpression>().binding;
+			ColumnBinding &right_column_binding = cond.right->Cast<BoundColumnRefExpression>().binding;
+			if (inner_bindings_set.find(right_column_binding) != inner_bindings_set.end() &&
+			    outer_bindings_set.find(left_column_binding) != outer_bindings_set.end()) {
+				inner_keys_set.insert(right_column_binding);
+			} else {
+				join_conditions_valid = false;
```

Comment:
I actually don't know when you would get here. Usually all of the comparison bindings come from the left or right bindings. I actually think it's safe to take out the whole for loop. You can assume the join condition by this point is valid. Otherwise a binding error will already have been thrown.